src/jdk.compiler/share/classes/com/sun/tools/javac/comp/TransTypes.java

```diff
@@ -539,6 +546,270 @@ public void visitLambda(JCLambda tree) {
+        //code to always generate a method handle to an accessible method
+        return tree.ownerAccessible;
+    }
+
+    /**
+     * Erasure destroys the implementation parameter subtype
+     * relationship for intersection types.
+     * Have similar problems for union types too.
+     */
+    boolean interfaceParameterIsIntersectionOrUnionType(JCMemberReference tree) {
+        List<Type> tl = tree.getDescriptorType(types).getParameterTypes();
+        for (; tl.nonEmpty(); tl = tl.tail) {
+            Type pt = tl.head;
+            if (isIntersectionOrUnionType(pt))
+                return true;
+        }
+        return false;
+    }
+
+    boolean isIntersectionOrUnionType(Type t) {
+        switch (t.getKind()) {
+            case INTERSECTION:
+            case UNION:
+                return true;
+            case TYPEVAR:
+                TypeVar tv = (TypeVar) t;
+                return isIntersectionOrUnionType(tv.getUpperBound());
+        }
+        return false;
+    }
+
+    private boolean isProtectedInSuperClassOfEnclosingClassInOtherPackage(Symbol targetReference,
+                                                                          Symbol currentClass) {
+        return ((targetReference.flags() & PROTECTED) != 0 &&
+                targetReference.packge() != currentClass.packge());
+    }
+
+    /**
+     * This method should be called only when target release <= 14
+     * where LambdaMetaFactory does not spin nestmate classes.
+     *
+     * This method should be removed when --release 14 is not supported.
+     */
+    boolean isPrivateInOtherClass(JCMemberReference tree) {
+        assert !target.runtimeUseNestAccess();
+        return  (tree.sym.flags() & PRIVATE) != 0 &&
+                !types.isSameType(
+                        types.erasure(tree.sym.enclClass().asType()),
+                        types.erasure(env.enclClass.sym.asType()));
+    }
+
+    /**
+     * Does this reference need to be converted to a lambda
+     * (i.e. var args need to be expanded or "super" is used)
+     */
+    boolean needsConversionToLambda(JCMemberReference tree) {
+        return interfaceParameterIsIntersectionOrUnionType(tree) ||
+                tree.hasKind(ReferenceKind.SUPER) ||
+                needsVarArgsConversion(tree) ||
+                isArrayOp(tree) ||
+                (!target.runtimeUseNestAccess() && isPrivateInOtherClass(tree)) ||
+                isProtectedInSuperClassOfEnclosingClassInOtherPackage(tree.sym, env.enclClass.sym) ||
+                !receiverAccessible(tree) ||
+                (tree.getMode() == ReferenceMode.NEW &&
+                        tree.kind != ReferenceKind.ARRAY_CTOR &&
+                        (tree.sym.owner.isDirectlyOrIndirectlyLocal() || tree.sym.owner.isInner()));
+    }
+
+    /**
+     * Converts a method reference which cannot be used directly into a lambda
+     */
+    private class MemberReferenceToLambda {
```

Comment:
I've considered adding a completely new step for this, but decided against it, as for now there is an assumption that `TransTypes` is the first translation step after `Flow`, and we run a scanner before it to collect usages of various features, as well as supertype/subtype ordering information (as `TransTypes` needs to be run in a certain order - e.g. supertypes first).